### PR TITLE
Implement ballerina native support for compression/decompression

### PR DESCRIPTION
**Pull request by NatashaWso2**
## Automation tests
 - Unit tests 
   > All the native functions implemented for compression and decompression have been tested using unit tests.

### CODE REVIEW

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            log.error("I/O Exception when processing files " + e.getMessage());
+        } finally {
+            try {
+                if (zin != null) {
+                    zin.close();
+                }
+            } catch (IOException e) {
+                log.debug("I/O Exception when closing the input stream ", e);
+                log.error("I/O Exception when closing the input stream " + e.getMessage());
+            }
```
Comment (anupama-pathirage) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: Shall we use try-with-resources Statement to avoid this close.

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/compression/natives.bal [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+public native function unzipFile (string dirPath, string destDir);
+
+@Description {value:"Function to compress a folder"}
+@Param {value:"dirPath: Path of the directory to be compressed"}
+@Param {value:"destDir: Destination directory to place the compressed file"}
+public native function zipFile (string dirPath, string destDir);
+
+@Description {value:"Function to compress bytes"}
+@Param {value:"dirPath: Path of the directory to be compressed"}
+@Return {value:"compressed bytes of the file"}
```
Comment (anupama-pathirage) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/compression/natives.bal]: It should be "Compressed bytes of the file"

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/ZipFile.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            File dir = new File(dirPath);
+            List<String> filesListInDir = CompressionUtils.populateFilesList(dir);
+
+            try {
+                fos = new FileOutputStream(destDir);
+            } catch (FileNotFoundException e) {
+                log.debug("File with the specified pathname does not exist", e);
+                log.error("File with the specified pathname does not exist : " + destDir);
+            }
+            zos = new ZipOutputStream(fos);
```
Comment (anupama-pathirage) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/ZipFile.java]: What if fos is null here?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        return pathValue;
+    }
+
+    /**
+     * Get files inside the zip file
+     * @param zipFilePath path of the zip file
+     * @return list of files available inside the zipped file
+     * @throws IOException
+     */
+    public ArrayList<String> getFilesInsideZip(String zipFilePath) throws IOException {
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java]: This method can be private

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            pathValue = path.toAbsolutePath().toString();
+        }
+        return pathValue;
+    }
+
+    /**
+     * Get files inside the zip file
+     * @param zipFilePath path of the zip file
+     * @return list of files available inside the zipped file
+     * @throws IOException
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java]: Shall we add description?

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+
+    /**
+     * Create the directory name.
+     *
+     * @param name name of the directory
+     * @return directory name
+     */
+    private static String dirpart(String name) {
```
Comment (SupunS) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: Better to rename this as "getDirectoryName".

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        int s = name.lastIndexOf(File.separatorChar);
```
Comment (SupunS) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: Do null check on variable "name"

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+     */
+    private static boolean mkdirs(File outdir, String path) {
+        File d = new File(outdir, path);
+        return !d.exists() && d.mkdirs();
+    }
+
+    @Override
+    public BValue[] execute(Context context) {
+        byte[] content = getBlobArgument(context, SRC_AS_BYTEARRAY_FIELD_INDEX);
+        String destDir = getStringArgument(context, DEST_PATH_FIELD_INDEX);
```
Comment (SupunS) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: destDir can be null.
Do null check in all places where a String is passed as an argument to the function.
Applicable for all the functions.

File: tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    compression:zipFile(dirPath, destDir);
+}
+
+function testUnzipBytes(blob content, string destDir) {
+    compression:unzipBytes(content, destDir);
+}
+
+function testZipToBytes(string dirPath) (blob) {
+    return compression:zipToBytes(dirPath);
+}
```
Comment (SupunS) [inline tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal]: Add test cases for passing null as the content/dirPath/destPath

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/compression/natives.bal (hunk 2 above)
Comment (NatashaWso2) [reply stdlib/ballerina-builtin/src/main/ballerina/ballerina/compression/natives.bal]: Fixed 5ea59f6

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/ZipFile.java (hunk 3 above)
Comment (NatashaWso2) [reply stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/ZipFile.java]: Fixed 5ea59f6

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java (hunk 4 above)
Comment (NatashaWso2) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java]: Fixed with 5ea59f6

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java (hunk 5 above)
Comment (NatashaWso2) [reply tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java]: Fixed with 5ea59f6

File: tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal (hunk 9 above)
Comment (NatashaWso2) [reply tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal]: Added test cases to cover all scenarios when null is passed. Fixed with 5ea59f6

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java (hunk 6 above)
Comment (NatashaWso2) [reply stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: Fixed with 5ea59f6

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java (hunk 7 above)
Comment (NatashaWso2) [reply stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: Fixed with 5ea59f6

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+    }
+
+    @Test(description = "test unzipping/decompressing a byte array without the destination directory path",
+            expectedExceptions = BLangRuntimeException.class)
+    public void testUnzipBytesWithoutDestDirectory() throws IOException, URISyntaxException {
+        String dirPath = getAbsoluteFilePath("datafiles/compression/");
+        byte[] fileContentAsByteArray = Files.readAllBytes(new File(dirPath +
+                File.separator + "test.zip").toPath());
+        BString destDir = new BString(null);
```
Comment (SupunS) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/compression/CompressionTest.java]: I think the correct test would be to assign null to `destDir` (rather than `new BString(null)` ).
It would be better to write this test using ballerina. eg:
```ballerina
function testUnzipFile(string dirPath, string destDir, string folderToExtract) {
    compression:unzipFile(null, null, null);
}
```
Should do the same for all the four newly implemented methods, and for all input paramters separately.

File: tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal (hunk 9 above)
Comment (SupunS) [reply tests/ballerina-test/src/test/resources/test-src/nativeimpl/functions/compression.bal]: We may need to update the test cases. Please see my below comment.

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        try {
+            Path outdir = Paths.get(outputFolder);
+            zin = new ZipInputStream(new ByteArrayInputStream(fileContentAsByteArray));
+            ZipEntry entry;
+            String name, dir;
+            while ((entry = zin.getNextEntry()) != null) {
+                name = entry.getName();
+                if (!folderToUnzip.isEmpty() && name.startsWith(folderToUnzip)) {
+                    int index = name.lastIndexOf('/') + 1;
+                    // str.substring(str.lastIndexOf("/") + 1)
```
Comment (dakshika) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/compression/UnzipBytes.java]: remove commented lines
